File: typegraph/python/typegraph/host/host.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import os
import re
import sys
from typing import List

from typegraph.gen import imports
from typegraph.gen.types import Err, Ok, Result


def has_match(text: str, items: List[str]) -> bool:
    for pat_s in items:
        pattern = re.compile(pattern=pat_s)
        if bool(re.search(pattern, text)):
            return True
    return False


class HostImpl(imports.HostHost):
    def print(self, msg: str):
        print(msg)

    def eprint(self, msg: str):
        print(msg, file=sys.stderr)

    def expand_path(self, root: str, exclude: List[str]) -> Result[List[str], str]:
        try:
            result = []
            if os.path.isfile(root):
                result.append(root)
            for path, _, files in os.walk(root):
                for name in files:
                    file_path = os.path.join(path, name)
                    if not has_match(file_path, exclude):
                        result.append(file_path)

            return Ok(result)
        except Exception as e:
            return Err(str(e))

    def path_exists(self, path: str) -> Result[bool, str]:
        try:
            return Ok(os.path.isfile(path) or os.path.isdir(path))
        except Exception as e:
            return Err(str(e))

    def read_file(self, path: str) -> Result[bytes, str]:
        try:
            file = open(path, mode="rb")
            return Ok(file.read())
        except Exception as e:
            return Err(str(e))

    def write_file(self, path: str, data: bytes) -> Result[None, str]:
        try:
            dirname = os.path.dirname(path)
            os.makedirs(dirname, exist_ok=True)
            file = open(path, "wb")
            file.write(data)
            return Ok(None)
        except Exception as e:
            return Err(str(e))

    def get_cwd(self) -> Result[str, str]:
        try:
            return Ok(os.getcwd())
        except Exception as e:
            return Err(str(e))

File: typegraph/deno/sdk/src/host/host.mjs
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0
import * as fs from "node:fs";
import * as path from "node:path";
import * as process from "node:process";

function listAllFilesHelper(
  root,
  list,
  exclude,
) {
  const currStat = fs.statSync(root);
  if (!currStat.isDirectory()) {
    list.push(root);
    return;
  }
  search: for (const filename of fs.readdirSync(root)) {
    const fullPath = path.join(root, filename);
    if (exclude) {
      for (const pattern of exclude) {
        const reg = new RegExp(pattern);
        if (reg.test(fullPath)) continue search;
      }
    }
    listAllFilesHelper(fullPath, list, exclude);
  }
}

export function expandPath(
  root,
  exclude,
) {
  try {
    const ret = [];
    listAllFilesHelper(root, ret, exclude);
    return ret;
  } catch (err) {
    throw (err instanceof Error ? err.message : err);
  }
}

export function print(msg) {
  console.log(msg);
}

export function eprint(msg) {
  console.error(msg);
}

export function getCwd() {
  try {
    return process.cwd();
  } catch (err) {
    throw (err instanceof Error ? err.message : err);
  }
}

export function pathExists(filePath) {
  try {
    return fs.existsSync(filePath);
  } catch (err) {
    throw (err instanceof Error ? err.message : err);
  }
}

export function readFile(filePath) {
  try {
    const buffer = fs.readFileSync(filePath, null);
    return new Uint8Array(buffer);
  } catch (err) {
    throw (err instanceof Error ? err.message : err);
  }
}

export function writeFile(filePath, data) {
  try {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    void fs.writeFileSync(filePath, data, { flag: "w" });
  } catch (err) {
    throw (err instanceof Error ? err.message : err);
  }
}
